library/std/src/io/mod.rs

```diff
@@ -238,13 +238,47 @@
 //! contract. The implementation of many of these functions are subject to change over
 //! time and may call fewer or more syscalls/library functions.
 //!
+//! ## I/O Safety
+//!
+//! Rust follows an [I/O safety] discipline that is comparable to its memory safety discipline. This
+//! means that file descriptors can be *exclusively owned*. (Here, "file descriptor" is meant to
+//! subsume similar concepts that exist across a wide range of operating systems even if they might
+//! use a different name, such as "handle".) An exclusivley owned file descriptor is one that no
+//! other code is allowed to close, but the owner is allowed to close it any time. A type that owns
+//! its file descriptor should close it in its `drop` function. Types like [`File`] generally own
+//! their file descriptor. Similarly, file descriptors can be *borrowed*. This indicates that the
+//! file descriptor will not be closed for the lifetime of the borrow, but it does *not* imply any
+//! right to close this file descriptor, since it will likely be owned by someone else.
+//!
+//! The platform-specific parts of the Rust standard library expose types that reflect these
+//! concepts, see [`os::unix`] and [`os::windows`].
+//!
+//! To uphold I/O safety, it is crucial that no code closes file descriptors it does not own. In
+//! other words, a safe function that takes a regular integer, treats it as a file descriptor, and
+//! closes it, is *unsound*.
```

Comment:
> Without this stronger rule, it would be impossible to ever encapsulate a file description, because things like dup(rand()) would become permitted, and that would preclude some useful use cases.

Given that I can `dup` a file descriptor that I have merely borrowed, most of these usecases will already not work -- at least they won't work with `OwnedFd`.

The OS will reliably return EBADFD on any operation if the FD is closed, so there's not really any point in disallowing this that I can see.

Reply:
File description encapsulation requires encapsulating the `OwnedFd` and not implementing `AsFd` or handing out a `BorrowedFd`. So it takes some deliberate doing, but it is doable, as long as we disallow things like `dup(rand())` and `read(rand(), ...)` and so on.

The OS is documented to return `EBADF`, however any time you get that error on a closed fd, it only means you got lucky that no code, perhaps on another thread racing with you, did eg. `File::open` and reused the fd number.

Reply:
(let's merge this with the other thread about file object encapsulation)